From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Remi <remi.riu95@gmail.com>
Date: Sat, 5 Nov 2022 13:50:30 +0100
Subject: [PATCH] API mode Hybride


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index f1af93319d6b29d7bc709e7e273e8abe26c7b4af..f7c48cbc9a5d0348866048b8532e484c44ab2496 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -2457,6 +2457,17 @@ public final class Bukkit {
     }
     // Paper end
 
+    // Papyrus start
+    /**
+     * Gets whether the Server is in hybrid mode or not.
+     *
+     * @return true if the server is in hybrid mode, false otherwise
+     */
+    public static boolean getHybridMode() {
+        return server.getHybridMode();
+    }
+    // Papyrus end
+
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 8621dba7f10ab822b5b99ce0d05da58823605cb6..d1bb8cecd0734e2554dc82615466c2b6b9beaac1 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2136,4 +2136,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull org.bukkit.potion.PotionBrewer getPotionBrewer();
     // Paper end
+
+    // Papyrus start
+    /**
+     * Gets whether the Server is in hybrid mode or not.
+     *
+     * @return true if the server is in hybrid mode, false otherwise
+     */
+    boolean getHybridMode();
+    // Papyrus end
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index fe5a6928948bbc0ddae20997b98f7d5cfb3d7c04..e4924835975361650ac2c18d100e6aaff3b0b2c0 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2806,6 +2806,16 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @param canAlwaysOpenCommandBlocks whether this player can always open command blocks
      */
     void setCanAlwaysOpenCommandBlocks(boolean canAlwaysOpenCommandBlocks);
+
+    /**
+     * Returns whether this player is authenticated with the server.
+     * That is, if the player either used /login to authenticate,
+     * registered for the first time using /register,
+     * was auto-logged in (online ID or auth timeout),
+     * or was force-authenticated with the API.
+     * @return whether this player is authenticated
+     */
+    boolean isAuthenticated();
     // Papyrus end
 
     /**
diff --git a/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java b/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java
index c9f395064656dd0126410eb3c6e197baa450c063..67acde745bfa85f47e68c27adcce0b5bd5147245 100644
--- a/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityPotionEffectEvent.java
@@ -213,6 +213,10 @@ public class EntityPotionEffectEvent extends EntityEvent implements Cancellable
          * When an illusion illager makes himself disappear.
          */
         ILLUSION,
+        /**
+         * When a user is requested to login or register. No event will be triggered for this cause.
+         */
+        HYBRID_AUTH,
         /**
          * When all effects are removed due to a bucket of milk.
          */
