From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Remi <remi.riu95@gmail.com>
Date: Sat, 5 Nov 2022 15:01:21 +0100
Subject: [PATCH] Ajout ServerShutdownEvent


diff --git a/src/main/java/org/bukkit/event/server/ServerShutdownEvent.java b/src/main/java/org/bukkit/event/server/ServerShutdownEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..24ae7b23f69c4f9fbb5a89e4020e9ac67387f8fe
--- /dev/null
+++ b/src/main/java/org/bukkit/event/server/ServerShutdownEvent.java
@@ -0,0 +1,22 @@
+package org.bukkit.event.server;
+
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * This event is called when the server is about to shutdown.
+ */
+public class ServerShutdownEvent extends ServerEvent {
+    private static final HandlerList handlers = new HandlerList();
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
