From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Remi <remi.riu95@gmail.com>
Date: Thu, 10 Nov 2022 01:38:21 +0100
Subject: [PATCH] Les messages sont tous trusted


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index f058c19e2432824c39333955b795b23adedb07e9..14cd38ddd09c0d6b3e16eea63a346f5ca3984a3c 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1449,7 +1449,8 @@ public abstract class PlayerList {
     }
 
     public boolean verifyChatTrusted(PlayerChatMessage message) { // Paper - private -> public
-        return message.hasSignature() && !message.hasExpiredServer(Instant.now());
+        return true;
+        //return message.hasSignature() && !message.hasExpiredServer(Instant.now());
     }
 
     // CraftBukkit start
