From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Remi <remi.riu95@gmail.com>
Date: Thu, 10 Nov 2022 01:38:21 +0100
Subject: [PATCH] Les messages sont tous trusted


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index d1131375388608b1c3ec40892f0a54690fe547b3..4b86dd142a7e4702a24cb9d18e9dbe79c24a7dec 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1504,7 +1504,8 @@ public abstract class PlayerList {
     }
 
     public boolean verifyChatTrusted(PlayerChatMessage message, ChatSender profile) { // Paper - private -> public
-        return !message.hasExpiredServer(Instant.now()) && message.verify(profile);
+        return true;
+        //return !message.hasExpiredServer(Instant.now()) && message.verify(profile);
     }
 
     // CraftBukkit start
