From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Remi <remi.riu95@gmail.com>
Date: Thu, 10 Nov 2022 01:38:21 +0100
Subject: [PATCH] Les messages sont tous trusted


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 932b423bd3a0d75f576a9a88e73d8d66966d3c52..41626a77ce84cc40ffce1c1891378a05685cf626 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1495,7 +1495,8 @@ public abstract class PlayerList {
     }
 
     public boolean verifyChatTrusted(PlayerChatMessage message) { // Paper - private -> public
-        return message.hasSignature() && !message.hasExpiredServer(Instant.now());
+        return true;
+        //return message.hasSignature() && !message.hasExpiredServer(Instant.now());
     }
 
     // CraftBukkit start
