From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Remi <remi.riu95@gmail.com>
Date: Sat, 12 Nov 2022 01:32:48 +0100
Subject: [PATCH] PlayerCommand - On force le label en lowercase


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 917391cd6f5a6c75af053fd64b89e1a3160d7427..d5fba76b1df54ece4b45ab9739ab7f5e7c71cd67 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -710,6 +710,12 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
             this.getHandle().connection.disconnect(Component.translatable("multiplayer.disconnect.illegal_characters"), org.bukkit.event.player.PlayerKickEvent.Cause.ILLEGAL_CHARACTERS);
         } else {
             if (msg.startsWith("/")) {
+                // Papyrus start - PlayerCommand - On force le lowercase sur le label.
+                String[] spl = msg.split(" ");
+                String label = spl[0].toLowerCase();
+                msg = msg.replaceFirst(spl[0], label);
+                // Papyrus end
+
                 this.getHandle().connection.handleCommand(msg);
             } else {
                 // TODO text filtering
