From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Remi <remi.riu95@gmail.com>
Date: Mon, 17 Apr 2023 00:56:37 +0200
Subject: [PATCH] AMEND AVEC CommandBlockExecuteEvent


diff --git a/src/main/java/net/minecraft/world/level/BaseCommandBlock.java b/src/main/java/net/minecraft/world/level/BaseCommandBlock.java
index f13c37b253209ac5273597b5f7bbd052e766afc8..c9208b68fb0f4928fc55625793cdbbb9a8c9f9d1 100644
--- a/src/main/java/net/minecraft/world/level/BaseCommandBlock.java
+++ b/src/main/java/net/minecraft/world/level/BaseCommandBlock.java
@@ -128,7 +128,7 @@ public abstract class BaseCommandBlock implements CommandSource {
 
                         var event = new org.bukkit.event.block.CommandBlockExecuteEvent(
                             this.getBukkitSender(source),
-                            entity == null ? org.bukkit.craftbukkit.block.CraftBlock.at(source.getLevel(), new net.minecraft.core.BlockPos(source.getPosition())) : null,
+                            entity == null ? source.getBukkitLocation().getBlock() : null,
                             entity == null ? null : org.bukkit.craftbukkit.entity.CraftEntity.getEntity(minecraftserver.server, entity),
                             this.command
                         );
